src/lib/storage/lz4_segment.cpp

```diff
@@ -120,27 +140,222 @@ std::vector<pmr_string> LZ4Segment<pmr_string>::decompress() const {
+
+      /**
+        * In the case of the last block, the decompressed data is possibly smaller than _block_size (it is
+        * _last_block_size large). However, when using _last_block_size as size for the decompression buffer, the LZ4
+        * decompression fails. Therefore, the block is resized and shrunk afterwards.
+        */
+      if (block_index == end_block) {
+        decompressed_block.resize(_last_block_size);
+        decompressed_block.shrink_to_fit();
+
+        // Also set the offset for the end of the string.
+        block_end_offset = end_offset % _block_size;
+      }
+
+      // Extract the string from the current block via the offsets and append it to the result string strean.
+      const auto start_offset_it = decompressed_block.cbegin() + block_start_offset;
+      const auto end_offset_it = decompressed_block.cbegin() + block_end_offset;
+      auto partial_result = pmr_string{start_offset_it, end_offset_it};
+      result_string << partial_result;
+
+      // After the first iteration this is set to 0u since only the first block's start offset can't be equal to zero.
+      block_start_offset = 0u;
+    }
+    return pmr_string{result_string.str()};
+  }
+}
+
 template <typename T>
 std::shared_ptr<BaseSegment> LZ4Segment<T>::copy_using_allocator(const PolymorphicAllocator<size_t>& alloc) const {
-  auto new_compressed_data = pmr_vector<char>{_compressed_data, alloc};
+  auto new_lz4_blocks = pmr_vector<pmr_vector<char>>{alloc};
+  for (const auto& block : _lz4_blocks) {
+    new_lz4_blocks.emplace_back(pmr_vector<char>{block, alloc});
+  }
   auto new_null_values = pmr_vector<bool>{_null_values, alloc};
+  auto new_dictionary = pmr_vector<char>{_dictionary, alloc};
 
-  if (_offsets.has_value()) {
-    auto new_offsets = pmr_vector<size_t>(*_offsets, alloc);
-    return std::allocate_shared<LZ4Segment>(alloc, std::move(new_compressed_data), std::move(new_null_values),
-                                            std::move(new_offsets), _decompressed_size);
+  if (_string_offsets.has_value()) {
+    auto new_string_offsets = pmr_vector<size_t>(*_string_offsets, alloc);
+    return std::allocate_shared<LZ4Segment>(alloc, std::move(new_lz4_blocks), std::move(new_null_values),
+                                            std::move(new_dictionary), std::move(new_string_offsets), _block_size,
+                                            _last_block_size, _compressed_size);
   } else {
-    return std::allocate_shared<LZ4Segment>(alloc, std::move(new_compressed_data), std::move(new_null_values),
-                                            _decompressed_size);
+    return std::allocate_shared<LZ4Segment>(alloc, std::move(new_lz4_blocks), std::move(new_null_values),
+                                            std::move(new_dictionary), _block_size, _last_block_size, _compressed_size);
   }
 }
 
 template <typename T>
 size_t LZ4Segment<T>::estimate_memory_usage() const {
```

Comment:
Image a column with 100 chunks of default size of 100,000 tuples. Assuming this column is a completely empty string column, we are storing `100,000 * sizeof(size_t) * 100 = 80MB`.
Is that correct? Compared to the actual data that's kinda ... crazy.

Some (but less significant) goes for the `null_values` vector.

If I am not mistaken and the structures are that large: do you have any ideas what to do here?

Reply:
In the case of a totally empty string segment we could just make the offset vector empty, since they are never used. But as soon as only a single character is somewhere in the column, it is needed.

As for the null_values: the segment could also be made not nullable (same as the value segment) but that doesn't really solve the problem.

Reply:
As the purpose of this PR is to reduce the footprint of a segment as much as possible, it would be really nice to have both null values and the offsets only stored when necessary. Do you think you can cover this in this PR or is it a lot of work? I don't want to punish you with such things on the last mile. ;)

Reply:
Oh and one thing: why use `size_t` for the offets over `ChunkOffset`?

Reply:
Thinking about: leave the `size_t`. Actually, `uint32_t` should be sufficient, but we never know what cases we'll later see.
We should think about compression of offsets anyways at a later point in time.

Reply:
Please, consider using SIMDBP128 vector compression.